sdk/search/azure-search-documents/src/main/java/com/azure/search/documents/implementation/batching/SearchIndexingPublisher.java

```diff
@@ -224,21 +225,21 @@ private List<TryTrackingIndexAction<T>> createBatch() {
         return batchActions;
     }
 
-    private int fillFromQueue(List<TryTrackingIndexAction<T>> batch, List<TryTrackingIndexAction<T>> queue,
+    private static <T> int fillFromQueue(List<TryTrackingIndexAction<T>> batch, Deque<TryTrackingIndexAction<T>> queue,
         int requested, Set<String> duplicateKeyTracker) {
-        int offset = 0;
         int actionsAdded = 0;
-        int queueSize = queue.size();
 
-        while (actionsAdded < requested && offset < queueSize) {
-            TryTrackingIndexAction<T> potentialDocumentToAdd = queue.get(offset++ - actionsAdded);
+        Iterator<TryTrackingIndexAction<T>> iterator = queue.iterator();
+        while (actionsAdded < requested && iterator.hasNext()) {
+            TryTrackingIndexAction<T> potentialDocumentToAdd = iterator.next();
 
             if (duplicateKeyTracker.contains(potentialDocumentToAdd.getKey())) {
                 continue;
             }
 
             duplicateKeyTracker.add(potentialDocumentToAdd.getKey());
-            batch.add(queue.remove(offset - 1 - actionsAdded));
+            batch.add(potentialDocumentToAdd);
+            iterator.remove();
```

Comment:
Yeah, document operations are done by the document key and from discussions these operations are done in parallel so there aren't great guarantees on what happens if you did a document delete and document update on the same document. So our design was to prevent sending multiple operations on the same document in each batch.